### PR TITLE
TB-obj-parser

### PR DESCRIPTION
**Pull request by srahim457**
Added the json store function to the tools
There is a radio button object in the Hamiltonian tab that switches between uploading w90 or the uploading of pythTB.
load_data.py can parse the json and make a triqs lattice object.

### CODE REVIEW

**Comment by srahim457** [comment]
added the save_json.py to tools

**Review by the-hampel — APPROVED** [review]

File: load_data.py [hunk 1]
```diff
@@ -45,22 +46,105 @@ def load_config(contents, h5_filename, data):
...
+        units.append(tuple(i))
+
+    # extract the hoppings
+    vectors = []  # list of np arrays
+    temp_dict = {}  # this dict is to keep track of instances of conjugate pair vectors
+    hopping_dict = {}  # this one that will be sent to make a triqs lattice object
+
+    for i in hoppings:
+        # not correct atm
+        t = i[0]
```
Comment (the-hampel) [inline load_data.py]: This has to be fixed first. There can be off diagonal terms in the hoppings. Meaning that we have to store for each hopping a matrix norb x norb.

**Review by the-hampel — CHANGES_REQUESTED** [review]

**Review by the-hampel — APPROVED** [review]

File: tabs/tab1_callbacks.py [hunk 2]
```diff
@@ -484,7 +500,7 @@ def plot_Akw(tb_switch, akw_switch, colorscale, tb_data, akw_data, sigma_data):
 
         k_mesh = tb_data['k_mesh']
         fig.add_shape(type = 'line', x0=0, y0=0, x1=max(k_mesh['k_disc']), y1=0, line=dict(color='gray', width=0.8))
-
+        print(k_mesh['k_points_dash'])
```
Comment (the-hampel) [inline tabs/tab1_callbacks.py]: remove debug print out

File: tabs/tab1_callbacks.py [hunk 3]
```diff
@@ -129,9 +131,34 @@ def change_button(pythTB_style, pythTB_filename):
...
+        if 'k_mesh' in tb_data:
+            df = pd.DataFrame()
+            df['k'] = tb_data['k_mesh']['k_disc']
+            for band in range(len(tb_data['eps_nuk'])):
+                b = 'ε'+str(band) 
+                df[b] = tb_data['eps_nuk'][band]
+        #print(df)
+        csv_string = df.to_csv(index=False, encoding='utf-8', sep=',',
+                               float_format='%.6e')
+        print(csv_string)
```
Comment (the-hampel) [inline tabs/tab1_callbacks.py]: rm debug prin

File: tabs/tab1_callbacks.py [hunk 4]
```diff
@@ -129,9 +131,34 @@ def change_button(pythTB_style, pythTB_filename):
...
+    @app.callback(
+        Output(id('download-csv'), 'href'),
+        [Input(id('Akw-title'), 'n_clicks'),
+        Input(id('tb-data'), 'data')],
+        prevent_initial_call=True,
+    )
+    def func(n_clicks, tb_data):
+        df = pd.DataFrame({"no": [0,0,0,0], "data": [0,0,0,0], "you need to plot the figure before downloading the csv": [0,0,0,0]})
+        
+        #plot_data = list(zip(k_mesh['k_disc'],*tb_data['eps_nuk']))
```
Comment (the-hampel) [inline tabs/tab1_callbacks.py]: rm debug stuf

File: tabs/tab1_layout.py [hunk 5]
```diff
@@ -17,11 +18,15 @@ def layout(tb_data, tb_kslice_data, akw_data, ak0_data, sigma_data, loaded_data)
             make_dashboard(tb_data, tb_kslice_data, akw_data, ak0_data, sigma_data, loaded_data, 1),
             # column 2
             html.Div([
-                #html.H3('A(k,ω)', style={'textAlign': 'center'}),
-                dcc.Link(children=html.H3('A(k,ω)', style={'textAlign': 'center'}),
-                         href="https://en.wikipedia.org/wiki/Tight_binding",
-                         style={"color":"black", "text-decoration":"none"}
-                        ),
+                #html.H3('A(k,ω)', id=id('Akw-title'), style={'textAlign': 'center'}),
```
Comment (the-hampel) [inline tabs/tab1_layout.py]: can be removed

File: load_data.py [hunk 6]
```diff
@@ -87,14 +99,15 @@ def load_pythTB_json(contents):
     hopping_dict={}
     m_zero = np.zeros((norb, norb), dtype=complex)
     #on-site energy
+    print(site_energies)
```
Comment (the-hampel) [inline load_data.py]: rm debug print

File: load_data.py [hunk 7]
```diff
@@ -87,14 +99,15 @@ def load_pythTB_json(contents):
...
     hopping_dict[(0, 0, 0)] = np.eye(norb, dtype=complex) * site_energies
     #hoppings
     for hop, orb_from, orb_to, vector in hoppings:
         if tuple(vector) not in hopping_dict:
             hopping_dict[tuple(vector)] = m_zero.copy()
             # per default pythTB does not explicitly stores -R
             hopping_dict[tuple(-np.array(vector))] = m_zero.copy()
-
+            print(hop)
```
Comment (the-hampel) [inline load_data.py]: rm debug print

File: tools/calc_tb.py [hunk 8]
```diff
@@ -24,6 +24,8 @@
 from triqs.gf import BlockGf
 from triqs.gf import GfReFreq, MeshReFreq
 from triqs.utility.dichotomy import dichotomy
+#from solid_dmft import _get_tb_bands as getTB
+import solid_dmft.postprocessing as sdmft
```
Comment (phibeck) [inline tools/calc_tb.py]: Please remove both lines.

File: tools/calc_tb.py [hunk 9]
```diff
@@ -90,19 +91,24 @@ def calc_tb_bands(data, add_spin, add_local, k_mesh, fermi_slice, band_basis = F
...
             e_vecs = np.zeros(e_mat.shape, dtype=complex)
+            total_proj = np.zeros(np.shape(e_vecs[0]))
             for ik in range(np.shape(e_mat)[2]):
                 evals, e_vecs[:,:,ik] = np.linalg.eigh(e_mat[:,:,ik])
                 e_mat[:,:,ik] = np.zeros(e_mat[:,:,ik].shape)
                 np.fill_diagonal(e_mat[:,:,ik],evals)
+            for band in range(data['n_wf']):
+                for orb in projected_orbs:
+                    total_proj[band] += np.real(e_vecs[orb,band] * e_vecs[orb,band].conjugate())
+            #print(total_proj)
```
Comment (phibeck) [inline tools/calc_tb.py]: Please remove last line.

File: tabs/tab2_callbacks.py [hunk 10]
```diff
@@ -59,7 +59,13 @@ def calc_tb(tb_switch, click_tb, add_spin, dft_mu, n_k, k_points, tb_kslice_data
             # calculate Hamiltonian
             tb_kslice_data['e_mat_re'] = e_mat.real.tolist()
             tb_kslice_data['e_mat_im'] = e_mat.imag.tolist()
-            tb_kslice_data['eps_nuk'], evec_nuk = tb.get_tb_kslice(tbl, k_mesh, dft_mu)
+            #tb_kslice_data['eps_nuk'], evec_nuk = tb.get_tb_kslice(tbl, k_mesh, dft_mu)
```
Comment (phibeck) [inline tabs/tab2_callbacks.py]: remove last line

File: tabs/dashboard.py [hunk 11]
```diff
@@ -239,6 +274,15 @@ def make_dashboard(tb_data, tb_kslice_data, akw_data, ak0_data, sigma_data, load
...
                             ], id=id('band-basis-tooltip'), style={'padding': '5px 5px'}
                             ),
+                            html.Div([
+                                html.P('select orbitals: ', style={'display': 'block', 'text-align': 'left', 'vertical-align': 'center'}
+                                       ),
+                                dcc.Input(id=id('select-orbitals'),
+                                          type='text', 
+                                          value='0',
+                                          debounce=True,
+                                          placeholder='select orbitals Ex: 0,1,2'),
```
Comment (phibeck) [inline tabs/dashboard.py]: remove "select orbitals" (it's already written explicitly.

File: tools/calc_akw.py [hunk 12]
```diff
@@ -271,19 +268,32 @@ def dens(mu):
...
+        eps_min, eps_max = tb_data['eps_min_max']
+        eps_nuk = tb.fourier(k_array)
+        #TODO: there could be an edge case that is beyond the boundaries
+        sigma_hartree = Sigma(0.0).real
+        sigma_eig, _ = np.linalg.eigh(sigma_hartree)
+        #print(sigma_hartree, sigma_eig)
+        if sigma_eig[-1] > 0: eps_max+=sigma_eig[-1]
+        if sigma_eig[0] <  0: eps_min+=sigma_eig[0]
+
+    print(eps_min, eps_max)
```
Comment (phibeck) [inline tools/calc_akw.py]: Remove last line.

**Review by phibeck — APPROVED** [review]

**Review by the-hampel — APPROVED** [review]

**Review by phibeck — APPROVED** [review]

**Review by phibeck — APPROVED** [review]

**Pull request merged**